File: tests/test_blockchain.py
import unittest
import hashlib
import rsa
from blockchain import Blockchain  # Assuming your blockchain code is in a file named blockchain.py

class TestBlockchain(unittest.TestCase):

    def setUp(self):
        """Set up a new blockchain instance for testing."""
        self.blockchain = Blockchain()
        self.blockchain.register_user("test_user", "test_password")

    def test_user_registration(self):
        """Test user registration."""
        self.assertIn("test_user", self.blockchain.users)
        hashed_password = hashlib.sha256("test_password".encode()).hexdigest()
        self.assertEqual(self.blockchain.users["test_user"], hashed_password)

    def test_user_authentication(self):
        """Test user authentication with correct credentials."""
        self.assertTrue(self.blockchain.authenticate_user("test_user", "test_password"))

    def test_user_authentication_failure(self):
        """Test user authentication with incorrect credentials."""
        self.assertFalse(self.blockchain.authenticate_user("test_user", "wrong_password"))

    def test_transaction_signing_and_verification(self):
        """Test transaction signing and verification."""
        private_key, public_key = rsa.newkeys(512)  # Generate RSA keys
        transaction = {"sender": "Alice", "recipient": "Bob", "amount": 50.0}
        signature = self.blockchain.sign_transaction(private_key, transaction)

        self.assertTrue(self.blockchain.verify_signature(public_key, signature, transaction))

    def test_invalid_signature(self):
        """Test verification of an invalid signature."""
        private_key, public_key = rsa.newkeys(512)  # Generate RSA keys
        transaction = {"sender": "Alice", "recipient": "Bob", "amount": 50.0}
        signature = self.blockchain.sign_transaction(private_key, transaction)

        # Modify the transaction to create an invalid signature
        transaction["amount"] = 100.0
        self.assertFalse(self.blockchain.verify_signature(public_key, signature, transaction))

    def test_add_transaction(self):
        """Test adding a valid transaction."""
        private_key, public_key = rsa.newkeys(512)  # Generate RSA keys
        transaction = {"sender": "Alice", "recipient": "Bob", "amount": 50.0}
        signature = self.blockchain.sign_transaction(private_key, transaction)

        self.blockchain.add_transaction(transaction["sender"], transaction["recipient"], transaction["amount"], signature)
        self.assertEqual(len(self.blockchain.current_transactions), 1)

    def test_mine_block(self):
        """Test mining a new block."""
        private_key, public_key = rsa.newkeys(512)  # Generate RSA keys
        transaction = {"sender": "Alice", "recipient": "Bob", "amount": 50.0}
        signature = self.blockchain.sign_transaction(private_key, transaction)

        self.blockchain.add_transaction(transaction["sender"], transaction["recipient"], transaction["amount"], signature)
        new_block = self.blockchain.mine_block()

        self.assertEqual(new_block.index, 1)
        self.assertEqual(len(new_block.transactions), 1)

    def test_validate_chain(self):
        """Test chain validation."""
        self.assertTrue(self.blockchain.validate_chain())

        # Tamper with the blockchain
        self.blockchain.chain[1].transactions[0]["amount"] = 100.0
        self.assertFalse(self.blockchain.validate_chain())

    def test_get_chain(self):
        """Test getting the blockchain."""
        chain = self.blockchain.get_chain()
        self.assertEqual(len(chain), 1)  # Only the genesis block should be present initially

        # Mine a block
        private_key, public_key = rsa.newkeys(512)  # Generate RSA keys
        transaction = {"sender": "Alice", "recipient": "Bob", "amount": 50.0}
        signature = self.blockchain.sign_transaction(private_key, transaction)
        self.blockchain.add_transaction(transaction["sender"], transaction["recipient"], transaction["amount"], signature)
        self.blockchain.mine_block()

        chain = self.blockchain.get_chain()
        self.assertEqual(len(chain), 2)  # Now there should be the genesis block and one mined block

if __name__ == "__main__":
    unittest.main()

File: tests/testBlockchain.js
const { expect } = require('chai');
const crypto = require('crypto');
const rsa = require('node-rsa');
const Blockchain = require('./blockchain'); // Assuming your blockchain code is in a file named blockchain.js

describe('Blockchain', function() {
    let blockchain;

    beforeEach(function() {
        blockchain = new Blockchain();
        blockchain.registerUser ("test_user", "test_password");
    });

    it('should register a user', function() {
        expect(blockchain.users).to.have.property("test_user");
        const hashedPassword = crypto.createHash('sha256').update("test_password").digest('hex');
        expect(blockchain.users["test_user"]).to.equal(hashedPassword);
    });

    it('should authenticate a user with correct credentials', function() {
        expect(blockchain.authenticateUser ("test_user", "test_password")).to.be.true;
    });

    it('should fail to authenticate a user with incorrect credentials', function() {
        expect(blockchain.authenticateUser ("test_user", "wrong_password")).to.be.false;
    });

    it('should sign and verify a transaction', function() {
        const key = new rsa({b: 512});
        const privateKey = key.exportKey('private');
        const publicKey = key.exportKey('public');
        const transaction = { sender: "Alice", recipient: "Bob", amount: 50.0 };
        const signature = blockchain.signTransaction(privateKey, transaction);

        expect(blockchain.verifySignature(publicKey, signature, transaction)).to.be.true;
    });

    it('should fail to verify an invalid signature', function() {
        const key = new rsa({b: 512});
        const privateKey = key.exportKey('private');
        const publicKey = key.exportKey('public');
        const transaction = { sender: "Alice", recipient: "Bob", amount: 50.0 };
        const signature = blockchain.signTransaction(privateKey, transaction);

        // Modify the transaction to create an invalid signature
        transaction.amount = 100.0;
        expect(blockchain.verifySignature(publicKey, signature, transaction)).to.be.false;
    });

    it('should add a valid transaction', function() {
        const key = new rsa({b: 512});
        const privateKey = key.exportKey('private');
        const publicKey = key.exportKey('public');
        const transaction = { sender: "Alice", recipient: "Bob", amount: 50.0 };
        const signature = blockchain.signTransaction(privateKey, transaction);

        blockchain.addTransaction(transaction.sender, transaction.recipient, transaction.amount, signature);
        expect(blockchain.currentTransactions).to.have.lengthOf(1);
    });

    it('should mine a new block', function() {
        const key = new rsa({b: 512});
        const privateKey = key.exportKey('private');
        const publicKey = key.exportKey('public');
        const transaction = { sender: "Alice", recipient: "Bob", amount: 50.0 };
        const signature = blockchain.signTransaction(privateKey, transaction);

        blockchain.addTransaction(transaction.sender, transaction.recipient, transaction.amount, signature);
        const newBlock = blockchain.mineBlock();

        expect(newBlock.index).to.equal(1);
        expect(newBlock.transactions).to.have.lengthOf(1);
    });

    it('should validate the blockchain', function() {
        expect(blockchain.validateChain()).to.be.true;

        // Tamper with the blockchain
        blockchain.chain[1].transactions[0].amount = 100.0;
        expect(blockchain.validateChain()).to.be.false;
    });

    it('should get the blockchain', function() {
        const chain = blockchain.getChain();
        expect(chain).to.have.lengthOf(1); // Only the genesis block should be present initially

        // Mine a block
        const key = new rsa({b: 512});
        const privateKey = key.exportKey('private');
        const publicKey = key.exportKey('public');
        const transaction = { sender: "Alice", recipient: "Bob", amount: 50.0 };
        const signature = blockchain.signTransaction(privateKey, transaction);
        blockchain.addTransaction(transaction.sender, transaction.recipient, transaction.amount, signature);
        blockchain.mineBlock();

        const updatedChain = blockchain.getChain();
        expect(updatedChain).to.have.lengthOf(2); // Now there should be the genesis block and one mined block
    });
});

// Run the tests
if (require.main === module) {
    require('mocha').run();
}
